File: 2520-using-a-robot-to-print-the-lexicographically-smallest-string/using-a-robot-to-print-the-lexicographically-smallest-string.go
func robotWithString(s string) string {
	// Hàm phụ để tìm ký tự nhỏ nhất còn lại dựa vào tần suất
	smallestRemainingChar := func(freq []int) byte {
		for i := 0; i < 26; i++ {
			if freq[i] > 0 {
				return byte('a' + i);
			}
		}
		return 'a'; // fallback
	}

	stack := []byte{};
	result := []byte{};
	freq := make([]int, 26);

	// Đếm tần suất ký tự trong chuỗi s
	for i := 0; i < len(s); i++ {
		freq[s[i]-'a']++;
	}

	// Duyệt từng ký tự trong chuỗi s
	for i := 0; i < len(s); i++ {
		ch := s[i];
		stack = append(stack, ch); // đưa ký tự vào stack (robot giữ trong t)
		freq[ch-'a']--;            // giảm số lượng ký tự còn lại

		// Nếu ký tự trên đỉnh stack nhỏ hơn hoặc bằng ký tự nhỏ nhất còn lại
		for len(stack) > 0 && stack[len(stack)-1] <= smallestRemainingChar(freq) {
			// Pop khỏi stack và thêm vào kết quả
			result = append(result, stack[len(stack)-1]);
			stack = stack[:len(stack)-1];
		}
	}

	// Ghi nốt các ký tự còn lại trong stack
	for len(stack) > 0 {
		result = append(result, stack[len(stack)-1]);
		stack = stack[:len(stack)-1];
	}

	return string(result);
}

File: 2520-using-a-robot-to-print-the-lexicographically-smallest-string/using-a-robot-to-print-the-lexicographically-smallest-string.java
class Solution {
    // Trả về ký tự nhỏ nhất còn lại dựa trên tần suất các ký tự
    private char smallestRemainingChar(int[] freq) {
        for (int i = 0; i < 26; i++) 
            if (freq[i] > 0) 
                return (char) ('a' + i);
        
        return 'a'; // default fallback
    }

    public String robotWithString(String s) {
        Stack<Character> stack = new Stack<>();
        StringBuilder result = new StringBuilder();
        int[] freq = new int[26]; // Đếm số lần xuất hiện mỗi chữ cái trong s

        // Đếm tần suất từng ký tự trong chuỗi s
        for (char ch : s.toCharArray())
            freq[ch - 'a']++;
        
        // Duyệt từng ký tự trong s
        for (char ch : s.toCharArray()) {
            // Đưa ký tự vào stack tạm t (do robot đang giữ t)
            stack.push(ch);
            freq[ch - 'a']--; // Giảm số lượng ký tự còn lại

            // Nếu ký tự trên đầu stack nhỏ hơn hoặc bằng ký tự nhỏ nhất còn lại trong s
            // thì đưa ra khỏi stack và ghi vào giấy (append vào kết quả)
            while (!stack.isEmpty() && stack.peek() <= smallestRemainingChar(freq)) 
                result.append(stack.pop());
        }

        // Nếu còn lại ký tự nào trong stack thì pop hết ra (tương đương ghi hết t còn lại)
        while (!stack.isEmpty()) 
            result.append(stack.pop());
        
        return result.toString(); // Trả về kết quả cuối cùng
    }
}